selfhost/repl.jakt

```diff
@@ -6,43 +6,86 @@ import utility { Span }
 import error { JaktError }
 import codegen { CodeGenerator, CodegenDebugInfo, ControlFlowState, AllowedControlExits }
 
-import extern "LibLine/Editor.h" {
-    namespace AK {
-        extern struct String {
-            function characters(this) -> raw c_char
-            function length(this) -> usize
-        }
-        extern struct Error {
-            function code(this) -> i32
-            function string_literal(this) -> String
-        }
-        extern struct ErrorOr<T> {
-            function is_error(this) -> bool
-            function release_value(mut this) -> T
-            function error(this) -> Error
-        }
-        extern struct Result<T> {
-            function is_error(this) -> bool
-            function release_value(mut this) -> T
-        }
-        extern struct NonnullRefPtr<T> {
-            function ptr(this) -> raw T
+import extern c "stdio.h" {
+    extern function fopen(pathname: raw c_char, mode: raw c_char) -> raw FILE
+    extern function fclose(stream: raw FILE) -> raw c_int
+    // Blame windows for this.
+    extern function fgets(s: raw c_char, n: usize, stream: raw FILE) -> raw c_char
+}
+
+function null<T>() -> raw T {
+    unsafe {
+        cpp {
+            "return nullptr;"
         }
     }
-    namespace Line {
-        extern struct Editor {
-            function try_create() -> AK::ErrorOr<AK::NonnullRefPtr<Editor> >
-            function get_line(mut this, anon prompt: raw c_char) -> AK::Result<AK::String>
+
+    abort()
+}
+
+function allocate<T>(count: usize) -> raw T {
+    unsafe {
+        cpp {
+            "return static_cast<T*>(malloc(count * sizeof(T)));"
         }
     }
+
+    abort()
+}
+
+enum LineResult {
+    Line(String)
+    Eof
 }
 
-function get_line(editor: &mut Line::Editor) throws {
-    let input_string = editor.get_line("> ".c_string()).release_value()
+struct Editor {
+    standard_input_file: raw FILE
+    line_pointer: raw c_char
+    prompt: String
 
-    mut builder = StringBuilder::create()
-    builder.append_c_string(input_string.characters())
-    return builder.to_string()
+    function create(prompt: String) throws -> Editor {
+        mut std_in = fopen("/dev/stdin".c_string(), "r".c_string())
+        if std_in == null<FILE>() {
+            eprintln("Could not open /dev/stdin for reading")
+            throw Error::from_errno(42)
+        }
+
+        let editor = Editor(
+            standard_input_file: std_in
+            line_pointer: allocate<c_char>(count: 4096)
```

Comment:
Did we not carry ByteBuffer over from AK?

Alternatively, is there like an "unsafe" way to get the .data() of a Jakt::Array of bytes?

Reply:
we did not, and there is not.